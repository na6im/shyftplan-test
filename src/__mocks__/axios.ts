import axios from 'axios';

const mockAxios = jest.genMockFromModule<typeof axios>('axios');
mockAxios.create = jest.fn(() => mockAxios) as any;

export default mockAxios;
