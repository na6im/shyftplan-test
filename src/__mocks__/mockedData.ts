export const eventsData = {
  items: [
    {
      id: 1,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-08-01T06:00:00.000Z',
      endsAt: '2020-08-01T11:00:00.000Z',
    },
    {
      id: 48,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-09-10T06:00:00.000Z',
      endsAt: '2020-09-10T14:00:00.000Z',
    },
    {
      id: 61,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-12-24T10:00:00.000Z',
      endsAt: '2020-12-24T18:00:00.000Z',
    },
    {
      id: 95,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-06-09T09:00:00.000Z',
      endsAt: '2020-06-09T13:00:00.000Z',
    },
    {
      id: 106,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-04-22T07:00:00.000Z',
      endsAt: '2020-04-22T13:00:00.000Z',
    },
    {
      id: 134,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-01-24T10:00:00.000Z',
      endsAt: '2020-01-24T15:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-10-06T07:00:00.000Z',
      endsAt: '2020-10-06T13:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-07-14T10:00:00.000Z',
      endsAt: '2020-07-14T14:00:00.000Z',
    },
    {
      id: 195,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-11-10T07:00:00.000Z',
      endsAt: '2020-11-10T13:00:00.000Z',
    },
    {
      id: 239,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-02-05T06:00:00.000Z',
      endsAt: '2020-02-05T13:00:00.000Z',
    },
  ],
  pagination: { offset: 0, limit: 10, count: 50 },
};

export const eventsDataNoMore = {
  items: [
    {
      id: 1,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-08-01T06:00:00.000Z',
      endsAt: '2020-08-01T11:00:00.000Z',
    },
    {
      id: 48,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-09-10T06:00:00.000Z',
      endsAt: '2020-09-10T14:00:00.000Z',
    },
    {
      id: 61,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-12-24T10:00:00.000Z',
      endsAt: '2020-12-24T18:00:00.000Z',
    },
    {
      id: 95,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-06-09T09:00:00.000Z',
      endsAt: '2020-06-09T13:00:00.000Z',
    },
    {
      id: 106,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-04-22T07:00:00.000Z',
      endsAt: '2020-04-22T13:00:00.000Z',
    },
    {
      id: 134,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-01-24T10:00:00.000Z',
      endsAt: '2020-01-24T15:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-10-06T07:00:00.000Z',
      endsAt: '2020-10-06T13:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-07-14T10:00:00.000Z',
      endsAt: '2020-07-14T14:00:00.000Z',
    },
    {
      id: 195,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-11-10T07:00:00.000Z',
      endsAt: '2020-11-10T13:00:00.000Z',
    },
    {
      id: 239,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-02-05T06:00:00.000Z',
      endsAt: '2020-02-05T13:00:00.000Z',
    },
  ],
  pagination: { offset: 0, limit: 10, count: 9 },
};

export const EventDetails = {
  id: 95,
  startsAt: '2020-06-09T09:00:00.000Z',
  endsAt: '2020-06-09T13:00:00.000Z',
  position: { name: 'Sports', color: 'lime', id: 45 },
  employees: [
    {
      firstName: 'Bettye',
      lastName: 'Mann',
      image: 'https://cdn.fakercloud.com/avatars/vivekprvr_128.jpg',
      id: 73,
    },
  ],
};
