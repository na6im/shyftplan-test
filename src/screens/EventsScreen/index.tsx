import React, { useState, useEffect } from 'react';
import { useInfiniteQuery } from 'react-query';
import { RouteComponentProps } from 'react-router-dom';
// @ts-ignore
import DatePicker from 'react-datepicker';
import api from '../../api';
import EventCard from '../../components/EventCard';
import { Event, Pagination } from '../../types';
import {
  AppContainer,
  LoadMoreButton,
  LoaderContainer,
  EventsContainer,
  FiltersContainer,
  Filter,
  ResetButton,
} from './styles';
import Loader from '../../components/Loader/Loader';
import logo from '../../assets/logo-only.png';
import 'react-datepicker/dist/react-datepicker.css';

interface Request {
  pagination: Pagination;
  items: Event[];
}
const EventsScreen: React.FC<RouteComponentProps> = () => {
  const [startsAt, setStartsAt] = useState(null);
  const [endsAt, setEndsAt] = useState(null);

  const {
    isLoading,
    isError,
    data,
    hasNextPage,
    fetchNextPage,
    isFetchingNextPage,
    refetch,
  } = useInfiniteQuery<Request>(
    'Events',
    (args) => api.getEvents(10, 0, args, startsAt, endsAt),
    {
      getNextPageParam: (lastPage) => {
        const {
          pagination: { offset, limit, count },
        } = lastPage;
        if (offset + limit < count) {
          return offset + 10;
        }
        return false;
      },
    }
  );

  useEffect(() => {
    refetch();
  }, [startsAt, endsAt]);

  const resetFilter = () => {
    setStartsAt(null);
    setEndsAt(null);
  };

  if (isLoading)
    return (
      <LoaderContainer>
        <Loader src={logo} />
      </LoaderContainer>
    );
  if (isError) return <LoaderContainer>Error</LoaderContainer>;

  return (
    <AppContainer>
      <FiltersContainer>
        <Filter>
          <div>
            <b>Starts At: </b>
          </div>
          <DatePicker
            selected={startsAt}
            onSelect={setStartsAt}
            onChange={setStartsAt}
          />
        </Filter>
        <Filter>
          <div>
            <b>Ends At: </b>
          </div>
          <DatePicker
            selected={endsAt}
            onSelect={setEndsAt}
            onChange={setEndsAt}
          />
        </Filter>
        <ResetButton onClick={resetFilter}> Reset </ResetButton>
      </FiltersContainer>
      <EventsContainer>
        {data?.pages?.map((group: any) =>
          group.items?.map((event: Event) => (
            <EventCard eventElement={event} key={event.id} />
          ))
        )}
      </EventsContainer>
      {hasNextPage && (
        <LoadMoreButton
          disabled={isFetchingNextPage}
          onClick={() => fetchNextPage()}
        >
          {' '}
          {isFetchingNextPage ? 'Loading...' : 'Load More'}{' '}
        </LoadMoreButton>
      )}
    </AppContainer>
  );
};

export default EventsScreen;
