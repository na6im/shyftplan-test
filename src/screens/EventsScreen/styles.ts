import styled from 'styled-components';

export const AppContainer = styled.div`
  width: 100vw;
  display: flex;
  align-items: center;
  flex-direction: column;
  margin: 2em;
`;
export const EventsContainer = styled.div`
  display: flex;
  width: 100%;
  flex-flow: row wrap;
`;

export const LoadMoreButton = styled.button`
  cursor: pointer;
  padding: 10px 20px;
  margin: auto;
  transition: all 500ms ease;
  border-radius: 5px;
  background-color: rgb(67, 135, 45);
  color: white;
  font-weight: bold;
  border: none;
  &:hover {
    transform: scale(1.1);
  }
  &:active {
    transform: scale(1.05);
  }
`;

export const LoaderContainer = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

export const FiltersContainer = styled.div`
    display: flex;
    flex-direction: row;
  align-items: center;
`
export const Filter = styled.div`
  display: flex;
  flex-direction: row;
  margin-right: 10px;
`

export const ResetButton = styled.button`
  cursor: pointer;
  padding: 5px 20px;
  margin: auto;
  transition: all 500ms ease;
  border-radius: 5px;
  background-color: rgb(67, 135, 45);
  color: white;
  font-weight: bold;

  &:hover {
    transform: scale(1.1);
  }
  &:active {
    transform: scale(1.05);
  }
`
