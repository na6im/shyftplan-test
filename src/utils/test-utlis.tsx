// test-utils.js
import React, { ComponentType, ReactChild, ReactChildren } from 'react';
import { render } from '@testing-library/react';
import { ToastProvider } from 'react-toast-notifications';
import { QueryClient, QueryClientProvider } from 'react-query';

interface AuxProps {
  children: ReactChild | ReactChildren;
}

const queryClient = new QueryClient();

function renderWithTheme(ui: React.ReactElement) {
  const Wrapper = ({ children }: AuxProps) => (
    <QueryClientProvider client={queryClient}>
      <ToastProvider>
        {children}
      </ToastProvider>
    </QueryClientProvider>
  );
  return {
    ...render(ui, { wrapper: Wrapper as ComponentType }),
  };
}

// override render method

export * from '@testing-library/react';

export { renderWithTheme };
