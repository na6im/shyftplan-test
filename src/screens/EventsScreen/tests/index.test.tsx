import React from 'react';
import { renderWithQuery, cleanup, screen } from '../../../utils/test-utlis';
import EventsScreen from '../index';
import { eventsData } from '../../../__mocks__/mockedData';
import { fireEvent } from '@testing-library/react';
import mockedAxios from '../../../__mocks__/axios';


describe('<EventsScreen />', () => {
  beforeEach(cleanup);
  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('should show loading then container', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: eventsData })
    );
    const { getByTestId } = renderWithQuery(<EventsScreen />);

    expect(getByTestId('isLoading')).toBeInTheDocument();
    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
  });

  it('should render events', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: eventsData })
    );
    renderWithQuery(<EventsScreen />);

    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
    const block = await screen.findAllByTestId('event');
    expect(block).toHaveLength(10);
  });

  it('should render loadMore', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: eventsData })
    );
    renderWithQuery(<EventsScreen />);

    const loadMoreButton = await screen.findByText(/Load More/i);
    expect(loadMoreButton).toBeInTheDocument();

    fireEvent.click(loadMoreButton);

    expect(mockedAxios.get).toBeCalled();
  });

  it('Should renderWithQuery and match the snapshot', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: eventsData,
      })
    );
    const {
      container: { firstChild },
    } = renderWithQuery(<EventsScreen />);

    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
    expect(firstChild).toMatchSnapshot();
  });
});
