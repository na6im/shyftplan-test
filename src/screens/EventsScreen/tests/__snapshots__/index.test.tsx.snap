// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventsScreen /> Should renderWithQuery and match the snapshot 1`] = `
<div
  class="sc-gtsrHT cVmcJU"
  data-testid="container"
>
  <div
    class="sc-jSFjdj inWfdf"
  >
    <div
      class="sc-gKAaRy kbMeAs"
    >
      <div>
        <b>
          Starts At: 
        </b>
      </div>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="sc-gKAaRy kbMeAs"
    >
      <div>
        <b>
          Ends At: 
        </b>
      </div>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <button
      class="sc-iCoGMd josGzm"
    >
       Reset 
    </button>
  </div>
  <div
    class="sc-dlnjwi foLcUm"
    data-testid="events"
  >
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Books
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        8/1/20, 7:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        8/1/20, 12:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Health
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        9/10/20, 7:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        9/10/20, 3:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        12/24/20, 11:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        12/24/20, 7:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        6/9/20, 10:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        6/9/20, 2:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Books
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        4/22/20, 8:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        4/22/20, 2:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Health
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        1/24/20, 11:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        1/24/20, 4:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        10/6/20, 8:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        10/6/20, 2:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        7/14/20, 11:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        7/14/20, 3:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        11/10/20, 8:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        11/10/20, 2:00 PM
      </div>
    </div>
    <div
      class="sc-bdnxRM cKHSxd"
      data-testid="event"
    >
      <div>
        <b>
          Name: 
        </b>
         
        Sports
      </div>
      <div>
        <b>
          Starts at: 
        </b>
        2/5/20, 7:00 AM
      </div>
      <div>
        <b>
          Ends at: 
        </b>
        2/5/20, 2:00 PM
      </div>
    </div>
  </div>
  <button
    class="sc-hKFxyN ehhQQQ"
  >
    Load More
  </button>
</div>
`;
